server/src/middleware/errorHandler.ts

```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2023 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+import { NextFunction, Request, Response } from 'express';
+import { ValidationError } from 'yup';
+
+export enum HttpCode {
+  BadRequest = 400,
+  Unauthorized = 401,
+  NotFound = 404,
+  InternalServerError = 500,
+  BadGateway = 502
+}
+
+export class CustomError extends Error {
+  public readonly statusCode: number;
+
+  constructor(message: string, statusCode: HttpCode) {
+    super(message);
+    this.statusCode = statusCode;
+    this.name = 'CustomError';
+  }
+}
+
+export class SisuError extends CustomError {
+  public readonly sisuErrorCode: number;
+
+  constructor(sisuErrorCode: number) {
+    super(`external API error: ${sisuErrorCode}`, HttpCode.BadGateway);
+    this.sisuErrorCode = sisuErrorCode;
+  }
+}
+
+export function ErrorHandler(err: unknown, req: Request, res: Response, next: NextFunction): void {
```

Comment:
The `NextFunction` doesn't seem to be get used. While there probably isn't currently a middleware that could do anything after this, it would feel like good practice to call it after processing the error. Correct me if I'm wrong!

Reply:
Reason why the `NextFunction` is declared here is that if not present, express will use its own default error handler. I don't know what kind of possible side-effects there is if the `NextFunction` is or is not used in the last middleware.

Reply:
Right, Express needs to make the decision based on the function arity. In that case, can we silence this lint somehow?

Reply:
`// eslint-disable-next-line no-unused-vars` may be one way to do it.

Reply:
eslint disabled 649b66633150df554734a031ce62fd34fd09718a